internals/approval_defs.py

```diff
@@ -153,7 +155,61 @@ def is_resolved(approval_values, field_id):
 
   # Any NOT_APPROVED value means that the review is no longer pending.
   for av in approval_values:
-    if av.state == review_models.Approval.NOT_APPROVED:
+    if av.state == Approval.NOT_APPROVED:
       return True
 
   return False
+
+def set_vote(
+    feature_id: int, gate_id: int, new_state: int, set_by_email: str) -> None:
+  """Add or update an approval value."""
+  if not Vote.is_valid_state(new_state):
+    raise ValueError('Invalid approval state')
+
+  now = datetime.datetime.now()
+  existing_list: list[Vote] = Vote.get_votes(
+      feature_id=feature_id, gate_id=gate_id, set_by=set_by_email)
+  if existing_list:
+    existing = existing_list[0]
+    existing.set_on = now
+    existing.state = new_state
+    existing.put()
+    return
+
+  new_vote = Vote(
+      feature_id=feature_id, gate_id=gate_id, state=new_state,
+      set_on=now, set_by=set_by_email)
+  new_vote.put()
+
+  # TODO(danielrsmith): As of today, there is only 1 gate per
+  # gate type and feature. Passing the gate ID will be required when adding
+  # UI functionality for multiple versions of the same stage/gate.
+  gates: list[Gate] = Gate.query(
+      Gate.feature_id == feature_id, Gate.gate_type == gate_id).fetch()
+  if len(gates) > 0:
+    update_gate_approval_state(gates[0])
+
+def _calc_gate_status(gate: Gate) -> int:
+  """Evaluates the current state that this gate should have."""
+  votes = Vote.get_votes(gate_id=gate.key.integer_id())
+  approvals = 0
+  for vote in votes:
+    if vote.state in (Vote.APPROVED, Vote.NA):
+      approvals += 1
+    elif vote.state == Vote.NOT_APPROVED:
+      return Vote.NOT_APPROVED
+  afd = APPROVAL_FIELDS_BY_ID[gate.gate_type]
+
+  if ((afd.rule == ONE_LGTM and approvals >= 1) or
+      (afd.rule == THREE_LGTM and approvals >= 3)):
+    return Vote.APPROVED
+  return Vote.NA
+
+def update_gate_approval_state(gate: Gate) -> int:
+  """Change the Gate state if it has changed."""
+  was_not_resolved = not gate.is_resolved()
+  gate.state = _calc_gate_status(gate)
+  gate.put()
+  if was_not_resolved and gate.is_resolved():
+    gate.clear_request()
```

Comment:
FYI, I'd like to stop deleting the review requests at some point soon.  That can come in a later PR, but just as a heads up: We should keep the review requests because they are part of the feature's history.  To decide which features to show in the "features pending my review" box, we will eventually query for Gate state rather than presence of review requests.

Alternatively (and maybe this is simpler), we could eliminate the concept of a Vote for a review request.  When the user pressed the Request review button, we could record a requester email and timestamp in the Gate and set the Gate state to indicate that a review is requested.  It seemed a little odd that a request is a special kind of vote.